### PR TITLE
Add enum value references from responses to extended references

### PR DESCRIPTION
**Pull request by bonnici**
* Updates the enhanced reference experimental feature so that we now also send references to enums from response objects in the reports.
* Adds some usage and config metrics related to the studio interop functionality.
* Changeset is covered in https://github.com/apollographql/router/pull/5331

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [x] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [x] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [x] Unit Tests
    - [ ] Integration Tests
    - [x] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

[ROUTER-375]: https://apollographql.atlassian.net/browse/ROUTER-375?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apollo-router/src/apollo_studio_interop/mod.rs
Comment (Geal) [inline apollo-router/src/apollo_studio_interop/mod.rs:234]: are all the tracing annotations still needed?

Comment (Geal) [reply apollo-router/src/apollo_studio_interop/mod.rs:234]: each of those annotations creates a new span, and there's some overhead for each created span even if not at a visible level

File: apollo-router/src/apollo_studio_interop/mod.rs
Comment (Geal) [inline apollo-router/src/apollo_studio_interop/mod.rs:313]:
```suggestion
            // Not using entry API here due to performance impact
            let enum_name_stats = match referenced_enums.get_mut(enum_name.as_str()) {
```

Comment (Geal) [reply apollo-router/src/apollo_studio_interop/mod.rs:313]: if we're not using the entry API, then there's no need to preallocate the key

File: apollo-router/src/apollo_studio_interop/mod.rs
Comment (Geal) [inline apollo-router/src/apollo_studio_interop/mod.rs:340]: this feels like the visitor created for demand control: https://github.com/apollographql/router/blob/dev/apollo-router/src/graphql/visitor.rs

@BrynCooke we should probably update this to move to the visitor after we merge (to make sure it won't block the release)

File: apollo-router/src/services/execution/service.rs
Comment (Geal) [inline apollo-router/src/services/execution/service.rs:361]: this should not be done twice. `format_response` happens twice because it first has to go through the filtered query (with fields removed due to authorization), then the client query, to get the expected shape by inserting nulls were appropriate. `extract_enums_from_response` should happen only once, and after those steps

**Review by Geal — CHANGES_REQUESTED** [review]
This looks fine overall, but 2 things should be done before merging:
- removing the tracing annotations that create debug level spans
- extracting enums only once, and after the 2 passes of response formatting have happened

File: apollo-router/src/apollo_studio_interop/mod.rs
Comment (bonnici) [reply apollo-router/src/apollo_studio_interop/mod.rs:234]: I thought they would be useful information to have in the trace view, but can remove if it's a performance issue. Are the spans created even if you're not running the router with `--log=trace`?

File: apollo-router/src/apollo_studio_interop/mod.rs
Comment (bonnici) [reply apollo-router/src/apollo_studio_interop/mod.rs:340]: I tried using a form of the visitor pattern a while ago but I can't remember why it didn't work for me. Maybe it was not available in router source before and just in federation, or maybe I'm using something that's not available from the visitor. In any case I'm happy to revisit this once all the changes are merged.

File: apollo-router/src/services/execution/service.rs
Comment (bonnici) [reply apollo-router/src/services/execution/service.rs:361]: So does the `query` contain all the enums that are also in `filtered_query` at this point in the code?

Comment (Geal) [reply apollo-router/src/services/execution/service.rs:361]: if an enum is annotated with `@authenticated` or other authorization directives and the client is not authorized, then the filtered query will not have the fields of that enum type. Do you want to measure what is obtained in query plan execution, or what is returned to the client?

Comment (bonnici) [reply apollo-router/src/services/execution/service.rs:361]: The reason for returning the enum usage is mainly so we can see if it's safe to remove an unused enum, so I think it's better to use the query plan execution.

**Review by o0Ignition0o — APPROVED** [review]

**Review by Geal — APPROVED** [review]

**Pull request merged**
